### PR TITLE
Cart

### PR DESCRIPTION
**Pull request by ula0156**
1) Simplified the cart class. 
2) Created CartsManager class and moved the rest of the methods from cart class here. 
3) Created cartsProvider, which handles communication with DB.
4) EF created several tables such as carts, sessions, they are part of the reservationModel.
5) Changed the way products are stored in cart. From dictionary of guid and int, transformed it to list of guid.
6) Changed the implementations of other classes which objects collaborate with cart, accordingly.

Verified that items when added to cart got reserved and appeared in ReservationDb as well as sessionId in sessionDb.

### CODE REVIEW

File: onlineShop.core/Data/Entities/Cart.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+using System.Linq;
+
+namespace onlineShop.core.Entities
+{
+    public class Cart
+    {
+        [Key]
+        public string SessionId { get; set; }
+
+        public string _Products {
```
Comment (simionliviumihai) [inline onlineShop.core/Data/Entities/Cart.cs]: Add some comments describing what is happening here.

File: onlineShop.core/Data/InMemory/InMemoryCartProvider.cs [hunk 2]
```diff
@@ -3,59 +3,65 @@
...
-        private Cart CreateNewCart()
-        {
-            return new Cart(
-                _productsProvider,
-                new ReservationsManager(
-                    _stocksProvider,
-                    _reservationsProvider));
-        }
-    }
+    //public class InMemoryCartProvider : ICartProvider
```
Comment (simionliviumihai) [inline onlineShop.core/Data/InMemory/InMemoryCartProvider.cs]: Why is all this code commented?

File: onlineShop.core/Managers/CartsManager.cs [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+{
+    public class CartsManager
+    {
+        private IReservationsProvider _reservationsProvider;
+        private IProductsProvider _productsProvider;
+        private IStocksProvider _stocksProvider;
+        private ReservationsManager _reservationsManager;
+        private ICartsProvider _cartsProvider;
+
+        public CartsManager(IStocksProvider stocksProvider, IProductsProvider productsProvider, IReservationsProvider reservationsProvider, ICartsProvider cartProvider)
```
Comment (simionliviumihai) [inline onlineShop.core/Managers/CartsManager.cs]: This line is too long, break it into multiple lines, one parameter per line

File: onlineShop.core/Managers/CartsManager.cs [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+        public Cart GetCartBySessionId(string sessionId)
+        {
+            // look for cart in CartDb
+            var cart = _cartsProvider.GetCartForSession(sessionId);
+
+            return cart;
+        }
+
+        //This method using ReservationManager is trying to reserve a product
+        // if product is reserved -> it aslo adding to the cart
```
Comment (simionliviumihai) [inline onlineShop.core/Managers/CartsManager.cs]: Misspelled also

File: onlineShop.core/Managers/CartsManager.cs [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        public double GetTotalWeight(Dictionary<Product, int> products)
+        {
+            double totalWeight = 0;
+            foreach (var item in products)
+            {
+                if (item.Key is PhysicalProduct)
+                {
+                    PhysicalProduct physicalItem = item.Key as PhysicalProduct;
+                    totalWeight += physicalItem.Size.Weight * item.Value;
+                    return totalWeight;
```
Comment (simionliviumihai) [inline onlineShop.core/Managers/CartsManager.cs]: This is wrong. Rethink what doesn't work

File: onlineShop.core/Managers/CartsManager.cs [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+                }
+            }
+
+            return 0;
+        }
+
+        //public void CompletePurchase(Cart cart)
```
Comment (simionliviumihai) [inline onlineShop.core/Managers/CartsManager.cs]: Why is this code commented? Also there are empty lines at the end

File: onlineShop.core/Specials/InactiveSessionsDetector.cs [hunk 7]
```diff
@@ -0,0 +1,35 @@
+﻿using onlineShop.Data.InMemory;
+using System;
+using System.Collections.Generic;
+using System.Timers;
+
+namespace onlineShop.core.Specials
+{
+    //public class InactiveSessionsDetector
```
Comment (simionliviumihai) [inline onlineShop.core/Specials/InactiveSessionsDetector.cs]: Add a TODO if you need to fix this commented code

File: onlineShopWeb/Views/Cart/DisplayCart.cshtml [hunk 8]
```diff
@@ -1,17 +1,40 @@
...
-            <td> @item.Value </td>            
-        </tr>
+        if (item.Key is Book)
+        {
+            <tr>
+                <td>@Html.ActionLink(item.Key.Name, "Details", "Books", new { id = item.Key.Id }, null)</td> 
+                <td> @item.Value </td> 
+                <td>@Html.ActionLink("Add copy", "AddToCart", "Cart", new { id = item.Key.Id }, null)</td> 
+                <td>@Html.ActionLink("Remove copy", "RemoveFromCart", "Cart", new { id = item.Key.Id }, null)</td>
+
```
Comment (simionliviumihai) [inline onlineShopWeb/Views/Cart/DisplayCart.cshtml]: You have empty lines here

**Review by simionliviumihai — CHANGES_REQUESTED** [review]

File: onlineShop.core/Data/Entities/Cart.cs (hunk 1 above)
Comment (ula0156) [reply onlineShop.core/Data/Entities/Cart.cs]: Done

File: onlineShop.core/Data/InMemory/InMemoryCartProvider.cs (hunk 2 above)
Comment (ula0156) [reply onlineShop.core/Data/InMemory/InMemoryCartProvider.cs]: Deleted it.

File: onlineShop.core/Managers/CartsManager.cs (hunk 5 above)
Comment (ula0156) [reply onlineShop.core/Managers/CartsManager.cs]: Fixed

File: onlineShop.core/Managers/CartsManager.cs (hunk 4 above)
Comment (ula0156) [reply onlineShop.core/Managers/CartsManager.cs]: Fixed.

File: onlineShop.core/Managers/CartsManager.cs (hunk 6 above)
Comment (ula0156) [reply onlineShop.core/Managers/CartsManager.cs]: Fixed

File: onlineShop.core/Specials/InactiveSessionsDetector.cs (hunk 7 above)
Comment (ula0156) [reply onlineShop.core/Specials/InactiveSessionsDetector.cs]: Added TODO

File: onlineShopWeb/Views/Cart/DisplayCart.cshtml (hunk 8 above)
Comment (ula0156) [reply onlineShopWeb/Views/Cart/DisplayCart.cshtml]: Done

File: onlineShop.core/Managers/CartsManager.cs (hunk 3 above)
Comment (ula0156) [reply onlineShop.core/Managers/CartsManager.cs]: Done

**Review by simionliviumihai — APPROVED** [review]

**Pull request closed without merging**